### DISCUSSION TITLE
Updating a relationship column with a conditional function based on other columns selections

### DISCUSSION THREAD

**Discussion by dora-daglarasar** [Budibase/budibase#7579] [Help] [answered]
I have 4 tables as T1, T2, T3, T4. 

On the Table T1, column X is in relationship T2 and column Y is in relationship with T3.
Based on the dropdown selections on X and Y columns, I want to assign a row to column Z from Table T4. In other words, I want column Z to update itself with a conditional function that looks up the X, Y values.

**Comment by melohagan** [comment, marked as answer]
Letters and number aren't fun to decipher! It's helpful if you can use real-life examples, a natural use case of some sort that is easier to follow.

These might help:

- https://docs.budibase.com/docs/formula#fields-from-relationships
- https://docs.budibase.com/docs/fetching-data#displaying-relational-data
- https://docs.budibase.com/docs/filter-by-relationships

**Reply by dora-daglarasar** [reply]
Hi @melohagan ,

Thank you for your reply. You're right indeed about letters and numbers. Sorry for that. 
I believe I cannot achieve what I want by using BudiBase DB but I've got to use PostgreSQL.

Here's the natural use case example;

I have 5 tables. 
1. Salaries
2. Product Types
3. Staff
4. Orders 
5. Salary Logs

Here are the columns;
-------------------  In Relationship ------------- Non-Related
An Order Row = { Staff; Product Types} ------- {Customer; Order Date; etc.}
A Salary Row = {Staff; Product Types} -------- {Amount($)}
A Salary Logs Row = {Staff; Order} ----------- {Amount($)}

- Each staff has a different salary amount based on the product type. (Ex: Mike, Product1, 40 $ ; Mike, Product2, 20 $ ; Ann, Product 2, 25$ ....)
- Say that an order is created, let's call it Order123. It has a staff and a product type.
- I want to automatically create a row into Salary Logs such as {Order123's Staff; Order123's Product Type, Salary Amount of that Staff on that Product Type}

I hope I explained it in an understandable way. I tried my best at least (ˆoˆ)

**Comment by melohagan** [comment]
Hey @dora-daglarasar thanks for the extra info!

The most convenient way to accomplish this might be through the use of automations: https://docs.budibase.com/docs/trigger

So you can have a **Trigger** that will automatically fire whenever a row from the Order table is created - the `Row Created` trigger.

Then you can add a **Create Row** [action](https://docs.budibase.com/docs/automation-actions) that will create the row in the Salary Logs table. 

You can use [bindings in automations](https://docs.budibase.com/docs/data-in-automations) to pass the information from the Order row that was created into the fields of the Salary Logs row that you want to create.

Hope that helps!

**Reply by dora-daglarasar** [reply]
Hi @melohagan,

Thank you for the information you've shared. It helped me to progress a lot.

Although, I am stuck with an automation problem. I am trying to achieve passing the triggering created row's Auto ID into my new created row column which is the automation result.

Let me give you an example with screenshots;

A Booking Row
<img width="1229" alt="Screen Shot 2022-09-05 at 14 28 52" src="https://user-images.githubusercontent.com/112421425/188438762-f94308ae-16a4-4c29-95b5-5ef3adff4582.png">
<img width="1218" alt="Screen Shot 2022-09-05 at 14 28 59" src="https://user-images.githubusercontent.com/112421425/188438771-d683ae32-fd29-4cba-98e8-830cf7d1c9f4.png">

Create Row Automation
<img width="451" alt="Screen Shot 2022-09-05 at 13 29 00" src="https://user-images.githubusercontent.com/112421425/188438807-19a3f556-80df-4845-8eb5-cf500770fc23.png">

Automation Test
<img width="340" alt="Screen Shot 2022-09-05 at 13 33 37" src="https://user-images.githubusercontent.com/112421425/188438834-0426157a-2eab-4c05-b056-aca396259498.png">

Created Salary Log
<img width="1208" alt="Screen Shot 2022-09-05 at 13 34 19" src="https://user-images.githubusercontent.com/112421425/188439536-651af0d8-cceb-449e-8fc7-e6bdc2d5b0f3.png">


I see that **Booking** and **BookingID** is not pushed to the Salary Log Row. 

What I'm trying to achieve is binding a Booking with the respectively created Salary Log. So that I can always access to a log's booking details.

On the other hand, when I try the Automation from published app, BookingID is passed but rest of the information is not passed.

Created Booking Row from the App
<img width="1373" alt="Screen Shot 2022-09-05 at 15 10 23" src="https://user-images.githubusercontent.com/112421425/188446331-4cc440f0-406e-4c5a-882c-847a9d4ea656.png">

Created by Salary Log Row by the Automation
<img width="1386" alt="Screen Shot 2022-09-05 at 15 10 40" src="https://user-images.githubusercontent.com/112421425/188446439-805149d6-c6e9-4c55-b992-9d97cc0ca0ce.png">

Is this caused by a bug that I caused? I'd appreciate if you please let me know if I should provide further information.

Many thanks in advance,
Kind Regards,
Dora

**Reply by melohagan** [reply]
I believe the auto ID binding is incorrect. The column name is actually "Auto ID" with a space. 
Columns with spaces are not directly supported (and there is an issue open here: https://github.com/Budibase/budibase/issues/5715)
However, you can get around this by using square brackets. So the binding would look as follows:

`{{ trigger.row[Auto ID] }}` *Note the space*

**Reply by dora-daglarasar** [reply]
Hi Dear @melohagan ,

Thank you for your reply. 
I actually did renamed the "Auto ID" as AutoID but for the sake of clarity, I have deleted my table and recreated it. I used [Auto ID] as you instructed.

However, I still couldn't pass all the fields with from the Automation Testing and from the Published App.


Below there are 2 rows and labelled. From the UI, I cannot pass the created Booking as an object. I can pass the Auto ID which is great. Staff and Salary Type is also cannot be pushed. However when I conduct Automation Test with the exact same Bindings, it works except the Auto ID and Booking Object which is sensible since test doesn't actually create the Booking row I assume.

Result Rows
<img width="1134" alt="Screen Shot 2022-09-07 at 11 43 17" src="https://user-images.githubusercontent.com/112421425/188835600-423d712d-9afa-4861-917c-fea7e5c788e9.png">

Automation Bindings
<img width="501" alt="Screen Shot 2022-09-07 at 11 54 04" src="https://user-images.githubusercontent.com/112421425/188835732-bc3a391c-4aa9-4c07-b74e-085e661984c7.png">

Booking Row
<img width="1228" alt="Screen Shot 2022-09-07 at 11 54 51" src="https://user-images.githubusercontent.com/112421425/188835897-182518cd-4eb5-47a2-87f9-7d157a502f0f.png">


I would also send you a big thanks for your interest and help. I really appreciate it. =))

**Reply by melohagan** [reply]
You're welcome! In this case you shouldn't have the dot before the square brackets. So try for example `{{ trigger.row[Auto ID] }}`
Note that there is no `.` after row. Let me know how that goes!

**Reply by dora-daglarasar** [reply]
Hi Dear @melohagan ,

Thank you for your reply.
I've tried without "." as {{ trigger.row[Auto ID] }} and unfortunately nothing changed.

By the way, with the previous dot included version {{ trigger.row.[Auto ID] }} I've managed to push the ID when I test from the published app. However cannot do it from the Automation Testing.

It's OK if Automation Testing doesn't push target object into new row. But the current blocker is; I cannot push the Staff and Salary Type from the published app although I can do it from the Automation Test.

**Screenshot from Published App**
<img width="1370" alt="Screen Shot 2022-09-08 at 12 18 57" src="https://user-images.githubusercontent.com/112421425/189085356-20c2b40d-f949-4cff-8e35-ca68674dcc90.png">


Below table is showing which of the fields I can push and I cannot with the exact same Automation setup.
<img width="565" alt="Screen Shot 2022-09-08 at 12 20 41" src="https://user-images.githubusercontent.com/112421425/189085523-866aa04e-895f-4726-9880-c40186f956b1.png">

So, I think there is a contradiction regardless of the syntax. Hope this explanation helps and maybe it'll contribute the amazing product you guys have. :))

Kind Regards,
-Dora

**Reply by melohagan** [reply]
Thanks Dora, would you mind sharing your app export to support@budibase.com please, and I can take a closer look?

**Reply by dora-daglarasar** [reply]
Hi @melohagan ,

Of course! I've sent the exported .txt doc to the support@budibase.com 

Thank you in advance!

**Reply by melohagan** [reply]
Thanks for the App Export Dora. I was able to figure out what was wrong - the test screen value uses an array of the relationship IDs instead of the array of objects that is used in the Design section. 
The workaround is to use the *pluck* handlebars expression, e.g.

![image](https://user-images.githubusercontent.com/101575380/189862204-37670b7b-2319-4b20-ae3d-ac8a0b0ad9cf.png)

I've raised a bug here: https://github.com/Budibase/budibase/issues/7750

As for the Auto ID, it should have twigged earlier, but the Auto ID is an autogenerated field, and thus should automatically increment and not need to be set.

**Reply by dora-daglarasar** [reply]
Hi,

Thank you! It's working and passing as you've instructed. Only this remains is to push the Booking object as {{ trigger.row }} itself. However, I can manage without it as well. :))

You've been so helpful, it's really laudable in my opinion. I might have new issues as I'm progressing now. Hope to chat later!

**Reply by melohagan** [reply]
You're welcome - glad to help!
